### PR TITLE
Use slicing kernel to copy distances inside NN Descent

### PR DESCRIPTION
**Pull request by jinsolp**
### Description
This make use of raft's slicing kernel within NN Descent build.
I found that my previous implementation was inefficient (merged in [this PR](https://github.com/rapidsai/raft/pull/2345)).

### Improvements
Time to call NN Descent `build()` with `return_distances=True` before and after using this kernel
| Dataset | Before |After|
| ------------- | ------------- |---|
| mnist (60000, 784)  | 1550ms  | 1020ms|
| sift (1M, 128)  | 11342ms  |5546ms|
| gist (1M, 960)  | 13508ms |9278ms|

### CODE REVIEW

File: cpp/include/raft/neighbors/detail/nn_descent.cuh [hunk 1]
```diff
@@ -1365,12 +1368,22 @@ void GNND<Data_t, Index_t, epilogue_op>::build(Data_t* data,
...
   if (return_distances) {
-    for (size_t i = 0; i < (size_t)nrow_; i++) {
-      raft::copy(output_distances + i * build_config_.output_graph_degree,
-                 graph_.h_dists.data_handle() + i * build_config_.node_degree,
-                 build_config_.output_graph_degree,
-                 raft::resource::get_cuda_stream(res));
-    }
+    auto graph_d_dists = raft::make_device_matrix<DistData_t, int64_t, raft::row_major>(
+      res, nrow_, build_config_.node_degree);
+    raft::copy(graph_d_dists.data_handle(),
```
Comment (lowener) [inline cpp/include/raft/neighbors/detail/nn_descent.cuh:1373]: Can this `raft::copy` be used: https://github.com/rapidsai/raft/blob/branch-24.08/cpp/include/raft/core/copy.cuh#L57? It supports mdspans.

Comment (jinsolp) [reply cpp/include/raft/neighbors/detail/nn_descent.cuh]: Thanks! didn't know we had that!

**Review by cjnolet — APPROVED** [review]
LGTM!

**Comment by jinsolp** [comment]
Okay now the types are a bit messy, because `graph_.h_dists` has index type `size_t`, and the mdspan `raft::copy` complains if `graph_d_dists` index type is not the same.

But then `slice` complains about meaningless comparison with 0 if the index type is not signed, which makes me have to set it to `int64_t` when getting the const view of `graph_d_dists`.

**Review by lowener — APPROVED** [review]

**Pull request merged**